File: implementations/sdk/java/src/test/java/com/spearsystems/lcp/ValidationCorpusTest.java
package com.spearsystems.lcp;

import org.junit.jupiter.api.Test;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;

import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;

final class ValidationCorpusTest {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    @Test
    void sharedValidationCorpusOutcomes() throws Exception {
        var root = Path.of("../../../");
        var validator = SchemaValidator.fromDirectory(root.resolve("schemas"));
        var corpus = MAPPER.readTree(Files.readString(root.resolve("test-vectors/sdk/validation-corpus.json")));
        var mismatches = new ArrayList<String>();
        for (var fixture : corpus.get("fixtures")) {
            var id = fixture.get("id").asText();
            var rule = fixture.get("rule").asText();
            var expectPass = fixture.get("expect").asText().equals("pass");
            var offer = fixture.get("offer");
            var isOffer = offer != null;
            var document = isOffer ? offer : fixture.get("envelope");
            var documentText = MAPPER.writeValueAsString(document);
            var passed = true;
            try {
                if (isOffer) {
                    validator.validateOffer(documentText);
                } else {
                    validator.validateEnvelope(documentText);
                }
            } catch (Exception e) {
                passed = false;
            }
            if (passed != expectPass) {
                mismatches.add(id + " (" + rule + "): expected " + (expectPass ? "pass" : "fail"));
            }
        }
        if (!mismatches.isEmpty()) {
            throw new AssertionError("Shared validation corpus mismatches: " + String.join("; ", mismatches));
        }
    }
}

File: implementations/sdk/kotlin/src/test/kotlin/com/spearsystems/lcp/ValidationCorpusTest.kt
package com.spearsystems.lcp

import java.nio.file.Path
import kotlin.test.Test
import tools.jackson.databind.ObjectMapper

class ValidationCorpusTest {
    @Test
    fun sharedValidationCorpusOutcomes() {
        val root = Path.of("../../../")
        val validator = LcpSchemaValidator.fromDirectory(root.resolve("schemas"))
        val mapper = ObjectMapper()
        val corpus = mapper.readTree(root.resolve("test-vectors/sdk/validation-corpus.json").toFile().readText())
        val mismatches = mutableListOf<String>()
        for (fixture in corpus.get("fixtures")) {
            val id = fixture.get("id").asText()
            val rule = fixture.get("rule").asText()
            val expectPass = fixture.get("expect").asText() == "pass"
            val offer = fixture.get("offer")
            val isOffer = offer != null
            val document = if (isOffer) offer else fixture.get("envelope")
            val documentText = mapper.writeValueAsString(document)
            val passed = try {
                if (isOffer) validator.validateOffer(documentText) else validator.validateEnvelope(documentText)
                true
            } catch (e: Exception) {
                false
            }
            if (passed != expectPass) {
                mismatches.add("$id ($rule): expected ${if (expectPass) "pass" else "fail"}")
            }
        }
        check(mismatches.isEmpty()) { "Shared validation corpus mismatches: ${mismatches.joinToString("; ")}" }
    }
}
